File: SRC/dlangt.f
      double           FUNCTION DLANGT( NORM, N, DL, D, DU );

*  -- LAPACK auxiliary routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      String             NORM;
      int                N;
      // ..
      // .. Array Arguments ..
      double             D( * ), DL( * ), DU( * );
      // ..

*  =====================================================================

      // .. Parameters ..
      double             ONE, ZERO;
      const              ONE = 1.0D+0, ZERO = 0.0D+0 ;
      // ..
      // .. Local Scalars ..
      int                I;
      double             ANORM, SCALE, SUM, TEMP;
      // ..
      // .. External Functions ..
      bool               LSAME, DISNAN;
      // EXTERNAL LSAME, DISNAN
      // ..
      // .. External Subroutines ..
      // EXTERNAL DLASSQ
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC ABS, SQRT
      // ..
      // .. Executable Statements ..

      if ( N.LE.0 ) {
         ANORM = ZERO
      } else if ( LSAME( NORM, 'M' ) ) {

         // Find max(abs(A(i,j))).

         ANORM = ABS( D( N ) )
         DO 10 I = 1, N - 1
            IF( ANORM.LT.ABS( DL( I ) ) .OR. DISNAN( ABS( DL( I ) ) ) ) ANORM = ABS(DL(I))             IF( ANORM.LT.ABS( D( I ) ) .OR. DISNAN( ABS( D( I ) ) ) ) ANORM = ABS(D(I))             IF( ANORM.LT.ABS( DU( I ) ) .OR. DISNAN (ABS( DU( I ) ) ) ) ANORM = ABS(DU(I))
         } // 10
      } else if ( LSAME( NORM, 'O' ) .OR. NORM.EQ.'1' ) {

         // Find norm1(A).

         if ( N.EQ.1 ) {
            ANORM = ABS( D( 1 ) )
         } else {
            ANORM = ABS( D( 1 ) )+ABS( DL( 1 ) )
            TEMP = ABS( D( N ) )+ABS( DU( N-1 ) )
            IF( ANORM .LT. TEMP .OR. DISNAN( TEMP ) ) ANORM = TEMP
            DO 20 I = 2, N - 1
               TEMP = ABS( D( I ) )+ABS( DL( I ) )+ABS( DU( I-1 ) )
               IF( ANORM .LT. TEMP .OR. DISNAN( TEMP ) ) ANORM = TEMP
            } // 20
         }
      } else if ( LSAME( NORM, 'I' ) ) {

         // Find normI(A).

         if ( N.EQ.1 ) {
            ANORM = ABS( D( 1 ) )
         } else {
            ANORM = ABS( D( 1 ) )+ABS( DU( 1 ) )
            TEMP = ABS( D( N ) )+ABS( DL( N-1 ) )
            IF( ANORM .LT. TEMP .OR. DISNAN( TEMP ) ) ANORM = TEMP
            DO 30 I = 2, N - 1
               TEMP = ABS( D( I ) )+ABS( DU( I ) )+ABS( DL( I-1 ) )
               IF( ANORM .LT. TEMP .OR. DISNAN( TEMP ) ) ANORM = TEMP
            } // 30
         }
      } else if ( ( LSAME( NORM, 'F' ) ) .OR. ( LSAME( NORM, 'E' ) ) ) {

         // Find normF(A).

         SCALE = ZERO
         SUM = ONE
         dlassq(N, D, 1, SCALE, SUM );
         if ( N.GT.1 ) {
            dlassq(N-1, DL, 1, SCALE, SUM );
            dlassq(N-1, DU, 1, SCALE, SUM );
         }
         ANORM = SCALE*SQRT( SUM )
      }

      DLANGT = ANORM
      RETURN

      // End of DLANGT

      }

File: lib/src/dlangt.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/install/lsame.dart';
import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/disnan.dart';
import 'package:dart_lapack/src/dlassq.dart';
import 'package:dart_lapack/src/matrix.dart';

double dlangt(
  final String NORM,
  final int N,
  final Array<double> DL_,
  final Array<double> D_,
  final Array<double> DU_,
) {
// -- LAPACK auxiliary routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final DL = DL_.having();
  final D = D_.having();
  final DU = DU_.having();
  const ONE = 1.0, ZERO = 0.0;
  int I;
  double ANORM = 0, TEMP;
  final SCALE = Box(0.0), SUM = Box(0.0);

  if (N <= 0) {
    ANORM = ZERO;
  } else if (lsame(NORM, 'M')) {
    // Find max(abs(A(i,j))).

    ANORM = D[N].abs();
    for (I = 1; I <= N - 1; I++) {
      if (ANORM < DL[I].abs() || disnan(DL[I].abs())) ANORM = DL[I].abs();
      if (ANORM < D[I].abs() || disnan(D[I].abs())) ANORM = D[I].abs();
      if (ANORM < DU[I].abs() || disnan(DU[I].abs())) ANORM = DU[I].abs();
    }
  } else if (lsame(NORM, 'O') || NORM == '1') {
    // Find norm1(A).

    if (N == 1) {
      ANORM = D[1].abs();
    } else {
      ANORM = D[1].abs() + DL[1].abs();
      TEMP = D[N].abs() + DU[N - 1].abs();
      if (ANORM < TEMP || disnan(TEMP)) ANORM = TEMP;
      for (I = 2; I <= N - 1; I++) {
        TEMP = D[I].abs() + DL[I].abs() + DU[I - 1].abs();
        if (ANORM < TEMP || disnan(TEMP)) ANORM = TEMP;
      }
    }
  } else if (lsame(NORM, 'I')) {
    // Find normI(A).

    if (N == 1) {
      ANORM = D[1].abs();
    } else {
      ANORM = D[1].abs() + DU[1].abs();
      TEMP = D[N].abs() + DL[N - 1].abs();
      if (ANORM < TEMP || disnan(TEMP)) ANORM = TEMP;
      for (I = 2; I <= N - 1; I++) {
        TEMP = D[I].abs() + DU[I].abs() + DL[I - 1].abs();
        if (ANORM < TEMP || disnan(TEMP)) ANORM = TEMP;
      }
    }
  } else if (lsame(NORM, 'F') || lsame(NORM, 'E')) {
    // Find normF(A).

    SCALE.value = ZERO;
    SUM.value = ONE;
    dlassq(N, D, 1, SCALE, SUM);
    if (N > 1) {
      dlassq(N - 1, DL, 1, SCALE, SUM);
      dlassq(N - 1, DU, 1, SCALE, SUM);
    }
    ANORM = SCALE.value * sqrt(SUM.value);
  }

  return ANORM;
}
